### PR TITLE
Add a tombstone operation to optimize multiple deletes

### LINKED ISSUE #31699: Optimize Remote Infinispan performance on removal of entry
### Description

The current implementation optimizes that a put and a removal cancel out each other so they will never touch the remote cache:

https://github.com/keycloak/keycloak/blob/9006218559496d5fb6fe98cd9760cad7f57b62e7/model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/remote/RemoteInfinispanKeycloakTransaction.java#L102-L106

This doesn't handle cases where the put it called once, and the remove is called multiple times. This is identical to the optimizations done in #28672

### Discussion

_No response_

### Motivation

Improve the performance of the remote cache as it has shown a regression in response time performance in https://github.com/keycloak/keycloak-benchmark/issues/901#issuecomment-2252501370 for cpuUsageForCredentialGrantsTest

### Details

I'll create a PR

### PR DESCRIPTION
**Pull request by ahus1**
This is a partial backport from KC26 to avoid a potential error in KC25

Closes #31699

### CODE REVIEW

**Comment by ahus1** [comment]
@mhajas - this is a partial backport of the change for KC26 which is IMHO relevant for KC25 as is prevents a corner case. Please review when you have the time. Thanks!

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanKeycloakTransaction.java [hunk 1]
```diff
@@ -122,10 +122,12 @@ public <K, V> void replace(Cache<K, V> cache, K key, V value, long lifespan, Tim
 
         Object taskKey = getTaskKey(cache, key);
         CacheTask current = tasks.get(taskKey);
-        if (current != null) {
+        if (current != null && current != TOMBSTONE && current.getOperation() != Operation.REMOVE) {
```
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanKeycloakTransaction.java:125]: @ahus1 I just realized we are changing the semantics of this method with this change. Previously when there was REMOVE and then REPLACE operation called on an existing entry it resulted with REMOVE of the entry. While now the result is REPLACE and DELETE is ignored. 

Is this change intended? Should we change the behavior in a patch release?

Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanKeycloakTransaction.java:125]: @mhajas - yes, this was intended. Once I added the 

```
throw new IllegalStateException("Can't replace entry: task " + current + " in progress for session");
```

some of the concurrency-tests started to fail. Looking at the (isolated) logic, a replace following a previous delete should do a replace, and not keep the original delete. I know this is the logic present in all previous releases, but it is really weird. Do we want to dive deeper on this and see which tests fail when an update is silently ignored when there was a previous remove?

Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanKeycloakTransaction.java:125]: I looked deeper into this, and it seems that `RootAuthenticationSessionAdapter` will call update even after it has called delete. Thank you for asking your questions here. 

I'm closing this backport, and will open a bug issue against main.
